src/gui/DatabaseOpenWidget.cpp

```diff
@@ -202,11 +202,31 @@ void DatabaseOpenWidget::openDatabase()
     QApplication::setOverrideCursor(QCursor(Qt::WaitCursor));
     m_ui->passwordFormFrame->setEnabled(false);
     QCoreApplication::processEvents();
-    bool ok = m_db->open(m_filename, databaseKey, &error, false);
+    KeePass2Reader::Status status = m_db->open(m_filename, databaseKey, &error, false);
     QApplication::restoreOverrideCursor();
     m_ui->passwordFormFrame->setEnabled(true);
 
-    if (ok) {
+    if (status == KeePass2Reader::Status::VersionWarn) {
+        QScopedPointer<QMessageBox> msgBox(new QMessageBox(this));
+        msgBox->setIcon(QMessageBox::Warning);
+        msgBox->setWindowTitle(tr("Database Version Mismatch"));
+        msgBox->setText(tr("The database you are trying to open was most likely\n"
+                           "created by a newer version of KeePassXC.\n\n"
+                           "You can try to open it anyway, but it may be incomplete\n"
+                           "and saving any changes may incur data loss.\n\n"
+                           "We recommend you update your KeePassXC installation."));
```

Comment:
I don't really understand this warning, when would a user ever see this? Why not just show a warning on open that the database may be incomplete due to unsupported KDBX version? What should the user do if they are not comfortable opening the database?

Reply:
This can be rather critical, so I would prefer if the user has to accept the case explicitly to be 100% sure they're aware of the (or some) danger. A simple warning at the top can be dismissed too easily and things could happen that trigger a save even if the user doesn't want to. KeePassXC ignores unknown fields on read and hence discards them on write, so you can lose data.

If you want a simple banner warning, we have to bring back the read-only mode.

Reply:
A shorter phrasing may be:
> The database version is not fully supported. Some data may not be read, and saving changes may incur data loss. Proceed anyway?

You could add the part about upgrading, but a newer minor DB version may not come from KPXC, and a newer KPXC version may not actually be available yet. The safest approach is to refuse to open it at all.

Reply:
Way too technical and way too vague. Users will be completely lost when they read that. "KeePassXC version too old" ist much clearer than "unsupported somehow for some reason". And "unsupported" is also generally a term used by tech people that scares non-technical folks.

Reply:
In this case, scaring them is probably a good thing. But you can adjust it as you see fit.

Reply:
Maybe this:

> You are trying to open a password database with a newer format than this version of KeePassXC supports. Continuing may result in loss of data or corruption. Would you like to continue?

Reply:
I don't like the word format there. It's too technical again. I tried to dumb it down as much as possible. Technical people will understand it anyway and other users are not thrown off entirely.

Reply:
IMO, if continuing may result in data loss, it's best not to give that option at all (unless really necessary, like with the different save modes). Users may just as easily click through a prompt without paying enough attention, and then wonder what happened to their data.

If you can detect all unsupported features when loading, then you can hard-fail only when such features are detected. Otherwise, why risk it? Or you can bring back read-only mode, as suggested.